File: compilers/craftinginterpreters/hlox/src/Lexer.hs
module Lexer where

import Data.Char (isDigit, isAlpha)
import Data.List (intercalate)
import Data.List.Split (splitOn)
import Data.Maybe (listToMaybe)
import Types

createToken :: Char -> String -> Int -> Either (ProgramError Char) TokenResult
createToken nextChar rest line
    | nextChar == '('                       = Right $ oneCharTokenWithoutRest LeftParen
    | nextChar == ')'                       = Right $ oneCharTokenWithoutRest RightParen
    | nextChar == '{'                       = Right $ oneCharTokenWithoutRest LeftBrace
    | nextChar == '}'                       = Right $ oneCharTokenWithoutRest RightBrace
    | nextChar == ':'                       = Right $ oneCharTokenWithoutRest Colon
    | nextChar == ','                       = Right $ oneCharTokenWithoutRest Comma
    | nextChar == '.'                       = Right $ oneCharTokenWithoutRest Dot
    | nextChar == '-'                       = Right $ oneCharTokenWithoutRest Minus
    | nextChar == '+'                       = Right $ oneCharTokenWithoutRest Plus
    | nextChar == ';'                       = Right $ oneCharTokenWithoutRest Semicolon
    | nextChar == '*'                       = Right $ oneCharTokenWithoutRest Star
    | nextChar == '!' && headRest == '='    = Right $ twoCharTokenWithoutRest BangEqual
    | nextChar == '!'                       = Right $ oneCharTokenWithoutRest Bang
    | nextChar == '=' && headRest== '='     = Right $ twoCharTokenWithoutRest EqualEqual
    | nextChar == '='                       = Right $ oneCharTokenWithoutRest Equal
    | nextChar == '<' && headRest == '='    = Right $ twoCharTokenWithoutRest LessEqual
    | nextChar == '<'                       = Right $ oneCharTokenWithoutRest Less
    | nextChar == '>' && headRest == '='    = Right $ twoCharTokenWithoutRest GreaterEqual
    | nextChar == '>'                       = Right $ oneCharTokenWithoutRest Greater
    | nextChar == '/' && headRest == '/'    = Right $ tokenWithRestAndLiteral Comment secondPartition firstPartition
    | nextChar == '/' && headRest == '*'    = Right $ tokenWithRestAndLiteral Comment commentRest comment
    | nextChar == '/'                       = Right $ oneCharTokenWithoutRest Slash
    | nextChar == '?'                       = Right $ oneCharTokenWithoutRest Question
    | nextChar == '\n'                      = createToken headRest tailRest (line + 1)
    | elem nextChar [' ', '\r', '\t']       = createToken headRest tailRest line
    | nextChar == '"'                       = createStringToken rest line
    | isDigit nextChar                      = createNumberToken nextChar rest line
    | isAlpha nextChar || nextChar == '_'   = Right $ createIdentifierOrKeywordToken nextChar rest line
    | otherwise                             = Left $ ProgramError (SourceCodeLocation Nothing line) "Unexpected character." rest
    where
        headRest = maybe '\0' id (listToMaybe rest)
        tailRest = if rest == [] then [] else tail rest
        secondPartition = getSecondElement partitions
        firstPartition = getFirstElement partitions
        partitions = splitOn ('\n':[]) rest
        (comment, commentRest) = getCommentAndRest tailRest
        getCommentAndRest :: String -> (String, String)
        getCommentAndRest content = (getFirstElement partition, getSecondElement partition)
            where partition = splitOn "*/" content
        getSecondElement :: [String] -> String
        getSecondElement (s:(s1:_)) = s
        getSecondElement _ = ""
        getFirstElement :: [String] -> String
        getFirstElement [] = ""
        getFirstElement (s:[]) = s
        tokenWithRestAndLiteral :: TokenType -> String -> String -> TokenResult
        tokenWithRestAndLiteral t r l = (createToken' t l, r, line + 1)
        oneCharTokenWithoutRest :: TokenType -> TokenResult
        oneCharTokenWithoutRest t = (createToken' t (nextChar : []), rest, line)
        twoCharTokenWithoutRest :: TokenType -> TokenResult
        twoCharTokenWithoutRest t = (createToken' t (nextChar : headRest : []), tailRest, line)
        createToken' :: TokenType -> String -> Token
        createToken' t s = Token t s $ SourceCodeLocation Nothing line

createNumberToken :: Char -> String -> Int -> Either (ProgramError Char) TokenResult
createNumberToken firstChar rest line = createNumberToken' [firstChar] rest
    where createNumberToken' :: String -> String -> Either (ProgramError Char) TokenResult
          createNumberToken' acc [] = maybeFinish acc []
          createNumberToken' acc (head:r) = if (isDigit head || head == '.') then createNumberToken' (acc ++ [head]) r else maybeFinish acc (head:r)
          numberToken :: Double -> TokenType
          numberToken n = TokenLiteral $ NumberLiteral n
          maybeFinish :: String -> String -> Either (ProgramError Char) TokenResult
          maybeFinish acc rest
              | last acc == '.'                       =
                    Left $ ProgramError (SourceCodeLocation Nothing line) "Invalid number." rest
              | (length $ filter (== '.') acc) > 1    =
                    Left $ ProgramError (SourceCodeLocation Nothing line) "Invalid number." rest
              | otherwise                             =
                    Right $ (Token (numberToken (read acc)) acc (SourceCodeLocation Nothing line), rest, line)

createStringToken :: String -> Int -> Either (ProgramError Char) TokenResult
createStringToken s line
    | elem '"' s    = Right $ (Token (stringToken string) (stringLiteral string) (SourceCodeLocation Nothing line), newRest, line + (breaklines string))
    | otherwise     = Left $ ProgramError (SourceCodeLocation Nothing (line+(breaklines s))) "Unterminated string." []
    where string = head partitions
          newRest = intercalate ['"'] (tail partitions)
          partitions = splitOn ('"':[]) s
          breaklines s = length $ filter (== '\n') s
          stringLiteral s = ('"':s) ++ "\""
          stringToken :: String -> TokenType
          stringToken s = TokenLiteral $ StringLiteral s

createIdentifierOrKeywordToken :: Char -> String -> Int -> TokenResult
createIdentifierOrKeywordToken nextChar input line = identifierOrKeywordToken word rest line
    where (word, rest) = extractWord (nextChar:input) []
          extractWord :: String -> String -> (String, String)
          extractWord [] acc = (acc, [])
          extractWord (c:rest) acc
              | isIdentifier c    = extractWord rest (acc ++ [c])
              | otherwise         = (acc, c:rest)
          isIdentifier :: Char -> Bool
          isIdentifier c = isDigit c || isAlpha c || '_' == c

identifierOrKeywordToken :: String -> String -> Int -> TokenResult
identifierOrKeywordToken word rest line
    | word == "and"   = createKeywordToken And word rest line
    | word == "class" = createKeywordToken Class word rest line
    | word == "else"  = createKeywordToken Else word rest line
    | word == "false" = createKeywordToken (TokenLiteral $ KeywordLiteral FalseKeyword) word rest line
    | word == "for"   = createKeywordToken For word rest line
    | word == "fun"   = createKeywordToken Fun word rest line
    | word == "if"    = createKeywordToken If word rest line
    | word == "nil"   = createKeywordToken (TokenLiteral $ KeywordLiteral NilKeyword) word rest line
    | word == "or"    = createKeywordToken Or word rest line
    | word == "print" = createKeywordToken Print word rest line
    | word == "return"= createKeywordToken Return word rest line
    | word == "super" = createKeywordToken Super word rest line
    | word == "this"  = createKeywordToken This word rest line
    | word == "true"  = createKeywordToken (TokenLiteral $ KeywordLiteral TrueKeyword) word rest line
    | word == "var"   = createKeywordToken Var word rest line
    | word == "while" = createKeywordToken While word rest line
    | otherwise       = createIdentifierToken word rest line

createKeywordToken :: TokenType -> String -> String -> Int -> TokenResult
createKeywordToken tokenType value rest line = (Token tokenType value (SourceCodeLocation Nothing line), rest, line)

createIdentifierToken :: String -> String -> Int -> TokenResult
createIdentifierToken value rest line =
    (Token (Identifier value) value (SourceCodeLocation Nothing line), rest, line)

scanTokens :: String -> ScanningResult
scanTokens s = scanTokens' s 1
    where
        scanTokens' :: String -> Int -> ScanningResult
        scanTokens' "" _ = Right []
        scanTokens' (h:t) l = (createToken h t l) >>= tokenResultToParseOutcome
        tokenResultToParseOutcome :: TokenResult -> ScanningResult
        tokenResultToParseOutcome (token, rest, line) = fmap ((:) token) $ scanTokens' rest line

File: compilers/craftinginterpreters/rloxi/src/lexer.rs
use crate::types::{Literal, ProgramError, SourceCodeLocation, Token, TokenType};
use std::str::FromStr;

pub struct Lexer {
    content: Vec<char>,
    current: usize,
    file: String,
    line: usize,
}

impl Lexer {
    pub fn new(s: String, file: String) -> Lexer {
        Lexer {
            file,
            content: s.chars().collect(),
            current: 0,
            line: 0,
        }
    }

    pub fn parse(&mut self) -> Result<Vec<Token>, Vec<ProgramError>> {
        let mut result = vec![];
        let mut errors = vec![];
        while self.current < self.content.len() {
            let lexem = self.content[self.current];
            let next = self.peek();
            let token = match (lexem, next) {
                ('(', _) => Some(self.create_token(TokenType::LeftParen, "(")),
                (')', _) => Some(self.create_token(TokenType::RightParen, ")")),
                ('{', _) => Some(self.create_token(TokenType::LeftBrace, "{")),
                ('}', _) => Some(self.create_token(TokenType::RightBrace, "}")),
                (':', _) => Some(self.create_token(TokenType::Colon, ":")),
                (',', _) => Some(self.create_token(TokenType::Comma, ",")),
                ('.', _) => Some(self.create_token(TokenType::Dot, ".")),
                ('-', _) => Some(self.create_token(TokenType::Minus, ")")),
                ('+', _) => Some(self.create_token(TokenType::Plus, "{")),
                (';', _) => Some(self.create_token(TokenType::Semicolon, ";")),
                ('/', Some('/')) => {
                    self.take_while(|c| c != '\n');
                    None
                },
                ('/', _) => Some(self.create_token(TokenType::Slash, "/")),
                ('*', _) => Some(self.create_token(TokenType::Star, "*")),
                ('?', _) => Some(self.create_token(TokenType::Question, "?")),
                ('!', Some('=')) => {
                    self.current += 1;
                    Some(self.create_token(TokenType::BangEqual, "!="))
                }
                ('!', _) => Some(self.create_token(TokenType::Bang, "!")),
                ('=', Some('=')) => {
                    self.current += 1;
                    Some(self.create_token(TokenType::EqualEqual, "=="))
                }
                ('=', _) => Some(self.create_token(TokenType::Equal, "=")),
                ('>', Some('=')) => {
                    self.current += 1;
                    Some(self.create_token(TokenType::GreaterEqual, ">="))
                }
                ('>', _) => Some(self.create_token(TokenType::Greater, ">")),
                ('<', Some('=')) => {
                    self.current += 1;
                    Some(self.create_token(TokenType::LessEqual, "<="))
                }
                ('<', _) => Some(self.create_token(TokenType::Less, "<")),
                ('\"', _) => {
                    let string_content = self.take_while(|s| s != '\"');
                    self.current += 1;
                    if self.current >= self.content.len() {
                        errors.push(self.create_error("Expected '\"', got end of string"));
                        None
                    } else {
                        self.current += 1;
                        let lexeme = format!("\"{}\"", string_content);
                        Some(self.create_token(
                            TokenType::TokenLiteral {
                                value: Literal::QuotedString(string_content),
                            },
                            lexeme.as_str(),
                        ))
                    }
                }
                (d, _) if d.is_digit(10) => {
                    let string_content = format!(
                        "{}{}",
                        d,
                        self.take_while(|s| s.is_digit(10) || s == '.' || s.is_alphabetic()),
                    );
                    match f32::from_str(string_content.as_str()) {
                        Ok(n) => Some(self.create_token(
                            TokenType::TokenLiteral {
                                value: Literal::Number(n),
                            },
                            &string_content,
                        )),
                        Err(_) => {
                            errors.push(self.create_error(&format!(
                                "Couldn't parse {} as number",
                                string_content
                            )));
                            None
                        }
                    }
                }
                (c, _) if c.is_alphabetic() || c == '_' => {
                    let string_content = format!(
                        "{}{}",
                        c,
                        self.take_while(|s| s.is_digit(10) || s.is_alphabetic() || c == '_'),
                    );
                    match string_content.as_str() {
                        "and" => Some(self.create_token(TokenType::And, "and")),
                        "class" => Some(self.create_token(TokenType::Class, "class")),
                        "else" => Some(self.create_token(TokenType::Else, "else")),
                        "fun" => Some(self.create_token(TokenType::Fun, "fun")),
                        "for" => Some(self.create_token(TokenType::For, "for")),
                        "break" => Some(self.create_token(TokenType::Break, "break")),
                        "if" => Some(self.create_token(TokenType::If, "if")),
                        "or" => Some(self.create_token(TokenType::Or, "or")),
                        "print" => Some(self.create_token(TokenType::Print, "print")),
                        "return" => Some(self.create_token(TokenType::Return, "return")),
                        "super" => Some(self.create_token(TokenType::Super, "super")),
                        "this" => Some(self.create_token(TokenType::This, "this")),
                        "var" => Some(self.create_token(TokenType::Var, "var")),
                        "while" => Some(self.create_token(TokenType::While, "while")),
                        _ => Some(self.create_token(
                            TokenType::Identifier {
                                name: string_content.clone(),
                            },
                            &string_content,
                        )),
                    }
                }
                ('\n', _) => {
                    self.line += 1;
                    None
                }
                ('\0', _) => Some(self.create_token(TokenType::EOF, "\0")),
                (a, _) if a.is_whitespace() => None,
                (c, _) => {
                    errors.push(self.create_error(&format!("Unexpected character {}", c)));
                    None
                }
            };
            if let Some(t) = token {
                result.push(t);
            }
            self.current += 1;
        }
        if errors.is_empty() {
            Ok(result)
        } else {
            Err(errors)
        }
    }

    fn create_error(&self, message: &str) -> ProgramError {
        ProgramError {
            location: self.get_current_location(),
            message: message.to_owned(),
        }
    }

    fn create_token(&self, token_type: TokenType, lexeme: &str) -> Token {
        Token {
            token_type,
            lexeme: lexeme.to_string(),
            location: self.get_current_location(),
        }
    }

    fn get_current_location(&self) -> SourceCodeLocation {
        SourceCodeLocation {
            file: self.file.clone(),
            line: self.line,
        }
    }

    fn peek(&self) -> Option<char> {
        let index = self.current + 1;
        self.content.get(index).cloned()
    }

    fn take_while<F: Fn(char) -> bool>(&mut self, f: F) -> String {
        let mut result = String::from("");
        while let Some(next) = self.peek() {
            if !f(next) {
                break;
            }
            self.current += 1;
            result.push(next);
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::Lexer;
    use crate::types::{Literal, ProgramError, SourceCodeLocation, Token, TokenType};

    #[test]
    fn test_lexer_with_no_error() {
    }

    #[test]
    fn test_lexer_with_unfinished_string() {
        let text = "var s = \"aasdfsadfsdfsadfasdfsdfaasdf";
        let mut lexer = Lexer::new(text.to_owned(), "file".to_owned());
        let expected = Err(vec![
            ProgramError {
                location: SourceCodeLocation {
                    file: "file".to_owned(),
                    line: 0,
                },
                message: "Expected \'\"\', got end of string".to_owned(),
            }
        ]);
        assert_eq!(lexer.parse(), expected);
    }

    #[test]
    fn test_lexer_with_unexpected_character() {
    }

    #[test]
    fn test_lexer_with_unparsable_number() {
    }

    #[test]
    fn test_lexer_with_more_than_one_error() {
    }
}
